In [4]:
import pandas as pd

df = pd.read_csv("C:\\Users\\sara\\Downloads\\fraude_bancario\\data\\creditcard.csv")

df.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0


In [5]:
df.shape

(284807, 31)

In [6]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [7]:
df['Class'].value_counts()


Class
0    284315
1       492
Name: count, dtype: int64

In [8]:
fraude = df['Class'].value_counts()[1]
total = df.shape[0]
print(f"Porcentaje de fraudes: {fraude / total * 100:.4f}%")


Porcentaje de fraudes: 0.1727%


In [9]:
x = df.drop(['Class'], axis=1)
y = df['Class']

In [10]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
x[['Amount', 'Time']] = scaler.fit_transform(x[['Amount', 'Time']])


In [11]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size=0.3, stratify=y, random_state=42
)


In [13]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)
x_train_bal, y_train_bal = smote.fit_resample(x_train, y_train)


De aca hacia abajo empezaremos a entrenar el modelo y nos basaremos en regresion logistica para que ya nos den los resultados.

In [22]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

lr = LogisticRegression()
lr.fit(x_train_bal, y_train_bal)

y_pred_lr = lr.predict(x_test)

print("Regresión Logística:\n")
print(classification_report(y_test, y_pred_lr))

print("Total: ")
print(confusion_matrix(y_test, y_pred_lr))


Regresión Logística:

              precision    recall  f1-score   support

           0       1.00      0.98      0.99     85295
           1       0.06      0.88      0.12       148

    accuracy                           0.98     85443
   macro avg       0.53      0.93      0.55     85443
weighted avg       1.00      0.98      0.99     85443

Total: 
[[83376  1919]
 [   18   130]]


Si lo pueden ejecutar pueden ver que puse el total de transacciones fraudulentas al final donde muestra que hubo 130 fraudulentas y aunque hay algunas no fraudulentas, por algunos datos las paso por sospechosas, la unica forma que tenemos de mejorar esto es cambiar el entrenamiento de modelos. Osea no utilizar regresion logistica, siempre porque no siempre nos dara buenos resultados o tan exactos como los necesitamos, ya que estas son transacciones bancarias.